### PR TITLE
Add callback interfaces for Kotlin

### LINKED ISSUE #352: Implement callback interfaces for Kotlin
This was #342 ; but now covers just the work in Kotlin.

[Callback interfaces][1]:

```
callback interface Filesystem {
    void write_text(string filepath, string text);
    string read_text(string filepath);
};

namespace my_lib {
    set_filesystem(Filesystem fs);
}
```

[1]: https://heycam.github.io/webidl/#dfn-callback-interface

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1922)

### PR DESCRIPTION
**Pull request by jhugman**
This PR implements the WebIDL `callback interface`, to setup Rust calling into Kotlin.

A callback interface describes a trait in Rust. Uniffi generates a corresponding interface in Kotlin for the client code to implement. The client code sends the implementation to rust, where Rust treats it as a `Box<dyn Trait>`. Any call to that trait is then sent back to Kotlin.

Related to #342 . (fixes for Kotlin)
Fixes #352 .

## Example Problem

We'd like to have a Rust trait to fetch a document which will drive experiments:

```rust
trait FetchClient {
    fn fetch(filepath: String) -> String
}

struct NimbusClient {
    client: Box<dyn FetchClient>,
}

impl NimbusClient {
    fn set_fetch_client(client: Box<dyn FetchClient>) {
        self.client = client;
    }

   fn update_experiments() {
        let json_string =  self.client.fetch("experiments.json");
        …
    }
}
```

Declared in the IDL:

```idl
callback interface FetchClient {
    string fetch(string filepath);
};

interface NimbusClient {
    set_fetch_client(FetchClient client);
    update_experiments();
};
```

And used in Kotlin:

```kotlin
class ConceptFetchClient(
    val baseURL: String
): FetchClient {
    override fun fetch(filepath: String): String {
       let url = URL(baseURL, filepath)
       …
    }
}

val nimbus = NimbusClient()
nimbus.setFetchClient(ConceptFetchClient(BASE_URL))
nimbus.updateExperiments() // uses ConceptFetchClient
```

## Implementation

This PR presents a strawfox implementation.

We sends JNA Callbacks to send a callback interface to Rust (a.k.a. `ForeignCallback`).

This implementation of the `ForeignCallback` is individual to the callback interface— on the kotlin side; in this case this is called: `CallbackInterfaceFetchClientFFI`.

As the `ConceptFetchClient` is sent from kotlin, it is placed in a kotlin `ConcurrentHandleMap`.

A Rust `impl` of corresponding `FetchClient` trait is generated which proxies all calls through the `ForeignCallback` to kotlin, i.e. `CallbackInterfaceFetchClientProxy` in Rust calls to `CallbackInterfaceFetchClientFFI` in kotlin.

The callback object i.e. the `ConceptFetchClient` is then found, and then methods are called from `CallbackInterfaceFetchClientFFI` to the client code `ConceptFetchClient`.

## Issues

 * No support for errors, or `Result`.
 * I'm a bit worried about threading with a second concurrent handlemap and the callback being able to call back into Rust.
 * I'm a bit worried about deadlocks in the same way as #197 (we're passing callbacks across the FFI boundary)
 * Testing of the Kotlin handlemap is non-existent; I have been unable to test if Drop was working properly with a `WeakReference`, but that should be more easily tested with handlemap testing.
 * I've cargo culted quite a lot of the viaduct code for setting callbacks, though I don't know if this is still the best way of doing things.
 * Other rusty things to do with `ffi_support` and saftey, I'm not super confident on.

I'm not sure about implementing `Send`, but it seemed the only way I could get a callback interface storable in an object.

### CODE REVIEW

**Comment by rfk** [comment]
Sorry I haven't had time to look at this in any detail, and I probably won't this week either :-(
Ping me if I don't seem to have gotten around to it by next week?

**Review by jhugman — COMMENTED** [review]
First pass with comments.

File: uniffi_bindgen/src/scaffolding.rs
Comment (rfk) [inline uniffi_bindgen/src/scaffolding.rs:106]: I don't think it should be possible to write `CallbackInterfaces` into a buffer, and it may be better to bail out here rather than generate code for it at all. (Also I think the closing paren on `Box::new(` may be missing here?)

File: uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs
Comment (rfk) [inline uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:7]: The comment here doesn't explain what the "generated proxy" is, it may be worth a sentence or two to do so (otherwise I don't know what the significance of it implementing these traits is)

File: uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs
Comment (rfk) [inline uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:31]: We should document here the reasons why it's safe to `impl Send` for this struct. Actually I'm surprised that it's not automatically `Send`, but I haven't finished following all the threads here yet.

File: uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs
Comment (rfk) [inline uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:50]: nit: I think it would be cleaner not to special-case a zero-length arguments list here, and just let the `for arg in meth.arguments()` loop execute zero times. The end result will be exactly the same (`RustBuffer::new()` calls `RustBuffer::from_vec` with an empty vec under the hood).

File: uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs
Comment (rfk) [inline uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:80]: Is it supported to return these back to the foreign-language code? If we don't have a good use-case for it I'd be tempted to make this an error, even if the implementation is obvious.

File: uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs
Comment (rfk) [inline uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:94]: As above, do we expect to ever want to read/write these in a buffer in practice? I don't have a specific failure mode in mind, it just makes me nervous...

File: uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs
Comment (rfk) [inline uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:100]: using a named const for the `0` here might make it easier to understand what this does at a glance; IIUC it would be something like `const IDX_CALLBACK_FREE = 0` or similar.

Comment (rfk) [reply uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:100]: The fact that we implement `Drop` here makes me very suspicious of writing these into a buffer etc, because we could serialize it to a buffer, drop it, then deserialize into a fresh instance with a handle that's been dropped.

File: uniffi_bindgen/src/interface/types.rs
Comment (rfk) [inline uniffi_bindgen/src/interface/types.rs:68]: nit: this comment feels incomplete - if it's not typically that, then what typically *is* it?

File: uniffi_bindgen/src/interface/mod.rs
Comment (rfk) [inline uniffi_bindgen/src/interface/mod.rs:1128]: nit: please say "callback interface" rather than just "interface" in these error strings.

File: uniffi_bindgen/src/interface/mod.rs
Comment (rfk) [inline uniffi_bindgen/src/interface/mod.rs:161]: nit: "an" -> "a"

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt
Comment (rfk) [inline uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:26]: In some future iteration, I guess this is where we'd signal an error back to the Rust code. Any thoughts on how that might work? (Or should we just forbid it for now?)

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt
Comment (rfk) [inline uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:72]: Oh, heh, it took me a bit of confusion to understand that `foreignCallback` here is the name of the thing you pass in to the constructor, but to know that you have to look at the implementation of `CallbackInternals`.

File: uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt
Comment (rfk) [reply uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt:55]: Right, good call. And things should be safe here once we've got a reference to the object out of the handlemap, because  the GC will keep it alive during the call.

File: uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt
Comment (rfk) [inline uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt:91]: As with the Rust side of things, I'm kind of dubious about `lift`/`read`/`write` here, since I don't think we can guarantee that the handle is still good.

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt
Comment (rfk) [inline uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:38]: What will actually happen to this exception if it does get thrown? Or equivalently, what will happen if the user's provided implementation method throws?

The JNA docs say:

> A callback should generally never throw an exception, since it doesn't necessarily have an encompassing
> Java environment to catch it. Any exceptions thrown will be passed to the default callback exception handler.

But it's not at all obvious to me what that actually means, or how the rust code will respond if allowed to continue running after this happens.

Should we try to catch exceptions and signal the error to the Rust code, maybe using an `out ExternError` argument like we do for calls from the foreign language into Rust?

**Review by rfk — COMMENTED** [review]
This is impressive stuff! I left a bunch of comments as I worked my way through it, but I don't think I came away with any substantial concerns. 

@jhugman are you looking to merge this and iterate, or are you after an initial round of feedback and plan to keep working on some of the outstanding things mentioned in the PR description?

The one thing I do think we need is some narrative docs about the high-level design of the system (like the kind of docs we *wish* we had about the serialization format). What do you think about starting a "design docs" subdirectory under `./docs` to contain such a thing?

File: examples/callbacks/src/callbacks.udl
Comment (jhugman) [reply examples/callbacks/src/callbacks.udl:25]: I started in `rondpoint` but quickly realised:

0. We couldn't land kotlin-only code and keep the `test_rondpoint.swift` and `test_rondpoint.py` going.
1. I wanted to write identifiers in English again.

File: uniffi/src/ffi/foreigncallbacks.rs
Comment (jhugman) [reply uniffi/src/ffi/foreigncallbacks.rs:35]: It certainly tidied things up on the kotlin side with a generic `CallbackInternals` class. This might be the basis of a rust `ForeignCallbackInternals`.

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt
Comment (jhugman) [reply uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:72]: I should add a `this.foreignCallback`.

File: uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt
Comment (jhugman) [reply uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt:55]: The `ConcurrentHandleMap` in Rust sounds like it has this same implementation and similar deadlock weakness.

File: uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt
Comment (jhugman) [reply uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt:91]: > I don't think we can guarantee that the handle is still good.

I'm not sure I understand why it's not true. I'll have to think about this.

**Comment by jhugman** [comment]
> are you looking to merge this and iterate, or are you after an initial round of feedback and plan to keep working on some of the outstanding things mentioned in the PR description?

AFAICT, the biggest unresolved issue left is trapping, logging and returning errors, which I'm not sure of the best path forward here.

I think a round of feedback would be useful and then spin out the missing pieces in to separate issues. i.e. land and iterate.

However, this isn't yet on the critical path, so it's going to be a while before it's ready to land.

**Review by travis79 — COMMENTED** [review]
This is awesome! I defer to @rfk's expertise on this, but I'm excited to try it out.

File: uniffi_bindgen/src/interface/mod.rs
Comment (jhugman) [reply uniffi_bindgen/src/interface/mod.rs:1128]: Yes. This was a bit of copy pasta.

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt
Comment (rfk) [reply uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:26]: Broadly agreed, the more symmetry we can have between `interface` and `callback interface` the better things are likely to be.

File: examples/callbacks/src/callbacks.udl
Comment (jhugman) [reply examples/callbacks/src/callbacks.udl:29]: I changed the names to not be `RoundTripper` and to try and reflect what language the called method is written in.

**Review by jhugman — COMMENTED** [review]
@rfk Thanks for feedback! I think this is ready for review now.

**Comment by rfk** [comment]
Great! I'll make sure this is in my review queue but probably won't get a chance to look at it until end of week or early next.

File: uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt
Comment (rfk) [reply uniffi_bindgen/src/bindings/kotlin/templates/CallbackInterfaceTemplate.kt:26]: Could you please add a brief `TODO` comment in the code here with a link to the follow-up issue? I think it's important that we remind our future selves that the `else` branch here is a temporary situation.

File: uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt
Comment (rfk) [inline uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt:37]: IIUC, this lock only protects concurrent access to the maps, and we release it before invoking the callback. So, it's possible to have multiple callbacks being dispatched at the same time. Accurate?

File: uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt
Comment (rfk) [reply uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt:98]: Revisiting this, I'm not sure I recall exactly what my concern was. I guess the other implicit link in the chain of argument here is that we don't ever `read` or `write` as part of some long-term serialization that might keep a reference to the handle around after it's been freed, we only do them as part of a transient shuffling of data from one side of the FFI to the other. So...I think I'm OK with this having concrete `read` and `write` implementations 👍

File: uniffi_bindgen/src/bindings/kotlin/templates/wrapper.kt
Comment (rfk) [inline uniffi_bindgen/src/bindings/kotlin/templates/wrapper.kt:64]: nano-nit: consider using a different name for the loop index variable here, e.g. `cb` or `callback` or something.

File: uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs
Comment (rfk) [reply uniffi_bindgen/src/templates/CallbackInterfaceTemplate.rs:36]: I tried commenting out this `unsafe impl Send` and `cargo test` still ran successfully to completion, which suggests to me that Rust is now automatically deriving a `Send` impl for us. Maybe a refactor has made it automatically `Send`?

What do you think about switching from `unsafe impl Send` to statically asserting that it is implemented? This would help remind us to re-check any assumptions here if we ever change the struct in a way that makes it no longer automatically `Send`.

(I'd also support taking a depending on the [`static_assertions`](https://docs.rs/static_assertions) crate for this, as it may be a bit awkward to express this as an expression suitable for use with `ffi_support::static_assert!`).

**Review by rfk — APPROVED** [review]
Sorry for the review delay here, the end of year kind of got away from me a bit.

This looks great! I left a bunch of small comments and nits and questions but nothing that should block landing this PR and iterating further in followups. I look forward to finding some time to work on a callback implementation for Python based on this :-)

One thing to consider is a brief "high-level overview of how callbacks work" document for our future reference. The module-level docstring of `foreigncallbacks.rs` might be a good place for this, absent a dedicated "design docs" area in the repo or manual.

File: uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt
Comment (jhugman) [reply uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt:37]: Yes, correct.

**Pull request merged**
